[CFM_58] API to upload an image /v1/image

0916dhkim opened the pull request:
## Changes
- Add `POST /v1/image` route handler.
- Add `S3Service` adapter class for interfacing with S3.

## Testing
1. Start localstack container:
    ```
    docker compose up
    ```
1. Start local SAM API
    ```
    sam local start-api --parameter-overrides ParameterKey=Environment,ParameterValue=local --docker-network cfm-network
    ```
1. Upload image (replace `<file-path>` with your actual image path like `"@/home/user/Downloads/sample.webp"`)
    ```
    curl --request POST \
      --url http://localhost:3000/v1/image \
      --header 'Content-Type: image/webp' \
      --data-binary "@<file-path>"
    ```
1. Check that it returns an URL and you can open the URL from your browser. Sample response:
    ```
    {"photoURL": "http://localhost:4566/fridge-report/dae7ad47-f1d5-4dc1-9214-03b2a68a6cb9.webp"}
    ```

0916dhkim commented:
@1jeanpaul1 I'm blocked on this feature. Here are two alternatives I considered:

1. [Localstack](https://localstack.cloud/) S3
  This looked like a great option because localstack is compatible with SAM ([integration](https://docs.localstack.cloud/integrations/aws-sam/)) and has S3 emulation. However, I could not use it for this project because lambda layer was a [paid](https://docs.localstack.cloud/aws/lambda/#lambda-layers) feature.
2. Saving files to local filesystem
  Saving files locally is tricky as well because `sam local start-api` runs code in a docker container. Saving files from `lambda_handler` saves them into the container, so we cannot access files from outside. It does not seem like there's a way to mount volume to SAM containers either.

1jeanpaul1 commented:
Could we use localstack to mock an s3 server in a separate port? And call that port when we're on dev mode?

https://medium.com/swlh/how-to-mock-aws-services-in-local-development-e231852e8a0f

0916dhkim commented:
@1jeanpaul1 Only using localstack for s3 and not for lambda sounds like a good approach to me. Let me see if I can run localstack s3 in a docker container like the post you shared :+1:

0916dhkim commented on docker-compose.yml:
```diff
@@ -1,12 +1,22 @@
 services:
-  dynamodb:
-    image: amazon/dynamodb-local:latest
+  localstack:
+    image: localstack/localstack:latest
```
Because Localstack includes DynamoDB, I removed the `dynamodb` container from the docker compose file.

0916dhkim commented on scripts/create_local_dynamodb_tables.py:
```diff
@@ -3,7 +3,7 @@
 from pathlib import Path
 
-ENDPOINT_URL = "http://localhost:8000"
+ENDPOINT_URL = "http://localhost:4566"
```
DynamoDB is now provided by Localstack whose gateway port is 4566.

0916dhkim commented on CommunityFridgeMapApi/tests/unit/test_image.py:
```diff
@@ -0,0 +1,34 @@
+from tests.storage_stub import StorageStub
+
+
+def test_upload(storage_stub: StorageStub):
+    blob = b'123123123'
```
Couple questions:
1. I did not use `unittest.TestCase` here because pytest fixture is not [supported](https://docs.pytest.org/en/7.1.x/how-to/unittest.html#pytest-features-in-unittest-testcase-subclasses) in unittest. Is this okay?
2. I've only written a happy-path case here. Let me know if we need to add any particular test cases.

1jeanpaul1 replied:
Yeah it's fine to not use unittest.TestCase as long as it still gets run

0916dhkim commented on CommunityFridgeMapApi/template.yaml:
```diff
@@ -7,6 +7,9 @@ Description: >
 
 # More info about Globals: https://github.com/awslabs/serverless-application-model/blob/master/docs/globals.rst
 Globals:
+  Api:
+    BinaryMediaTypes:
+      - image/webp
```
This is required to have a binary response or request.

1jeanpaul1 replied:
Does it need to be a global or can it be in `ImageFunction`?

0916dhkim replied:
To be honest, I'm not entirely sure how to make it specific to `ImageFunction`.

`BinaryMediaTypes` is a property of `AWS::Serverless::Api` resource type ([reference](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-api.html#sam-resource-api-properties)), and I could not find an equivalent property in `AWS::Serverless::Function` type.

0916dhkim commented on CommunityFridgeMapApi/functions/image/v1/app.py:
```diff
@@ -0,0 +1,45 @@
+    @staticmethod
+    def lambda_handler(event: dict, storage: Storage) -> dict:
+        bucket = "fridge-report"
+        key = storage.write(bucket, "webp", ImageHandler.get_binary_body_from_event(event))
```
I did not validate the uploaded data here. Do we want to check if the uploaded data is a valid webp image?

1jeanpaul1 replied:
Yes we should check if it's valid webp, return an error code and message if it's not

0916dhkim replied:
Added validation here f009311aaff1f17a5aa317aef2fb18e24c2f4a2a.

0916dhkim commented on CommunityFridgeMapApi/functions/image/v1/app.py:
```diff
@@ -0,0 +1,45 @@
+    @staticmethod
+    def get_binary_body_from_event(event: dict) -> bytes:
+        """Extract binary data from request body"""
+        assert event["isBase64Encoded"]
```
Making sure the request body is a base64 encoded binary data.
When user uploads a raw binary file, AWS lambda encodes the data in base64 before passing it to the `lambda_handler`. That's why we need to decode the data first before writing it in S3.

0916dhkim reviewed (commented):
It is a sizable PR, so I left some annotations to explain what's going on.
Also, I have some questions below :arrow_down:

1jeanpaul1 commented:
So I am able to make a request to the API and get a response: {"photoURL": "http://localhost:4566/fridge-report/2f155bb6-ad9b-40cf-a357-2ab71a8082f8.webp"}

But when I go to the link a file is returned that I am not able to open. The file size is also very small (49 bytes) compared to the image I uploaded so I assume that's part of the problem

1jeanpaul1 commented on CommunityFridgeMapApi/dependencies/python/db.py:
```diff
@@ -18,7 +18,7 @@ def get_ddb_connection(
     ddbclient = ""
     if env == "local":
-        ddbclient = boto3.client("dynamodb", endpoint_url="http://dynamodb:8000/")
+        ddbclient = boto3.client("dynamodb", endpoint_url="http://localstack:4566/")
```
The README needs to be updated if we're switching to using localstack for dynamodb.

Specifically I want to run: `aws dynamodb scan --table-name fridge --endpoint-url http://localstack:4566` but get this error: `Could not connect to the endpoint URL: "http://localstack:4566/"`

1jeanpaul1 replied:
The APIs work though which is good :). Am I pointing to the incorrect port when I try to do a `aws dynamodb scan`?

0916dhkim replied:
Yes, we need to point to http://**localhost**:4566 when running `aws dynamo scan` commands.
I've fixed port numbers in the README file.
We need to use **localstack** when accessing the localstack instance from within lambda, and **localhost** when accessing it from the host machine. This is because lambda containers are inside cfm-network.

1jeanpaul1 commented on CommunityFridgeMapApi/functions/image/v1/app.py:
```diff
@@ -0,0 +1,45 @@
+
+    @staticmethod
+    def lambda_handler(event: dict, storage: Storage) -> dict:
+        bucket = "fridge-report"
```
This API handles images from fridge and fridge-report. Should we put all of the images in the same folder? or use two different folders?

If we use two different folders should the client pass in a field to differentiate between fridge and fridge-report?

0916dhkim replied:
We can use a multipart form to pass-in a bucket name & binary file.
The spec only had single file upload, so that's why I implemented the lambda handler this way.

1jeanpaul1 replied:
can we name the bucket `CommunityFridgeMapImages`?

fridge-report makes it sounds like only fridge-report images will be uploaded here

0916dhkim replied:
Yep, fixed here e0086626d9f64e2dec63eddd9731bbd60f149af0

1jeanpaul1 commented on CommunityFridgeMapApi/functions/image/v1/app.py:
```diff
@@ -0,0 +1,45 @@
+        key = storage.write(bucket, "webp", ImageHandler.get_binary_body_from_event(event))
+        url = storage.generate_file_url(bucket, key)
+        return {
+            "statusCode": 200,
```
Are there any errors we need to catch? S3 has really high availability, but I wonder if s3 throws any exceptions when an image didn't upload correctly for some reason

0916dhkim replied:
I think that's a valid concern. Added some error handling in `S3Service` 39f4cf7d5b7f345a36e9cdd2111abf4ebdea471a.
The lambda handler should return 500 on error.

1jeanpaul1 commented on CommunityFridgeMapApi/dependencies/python/storage.py:
```diff
@@ -0,0 +1,85 @@
+        endpoint_url=endpoint_url,
+    )
+
+def translate_s3_url_for_client(url: str, env=os.getenv("Environment")) -> str:
```
Missing unit test for this function

0916dhkim replied:
Good catch :+1: 
Added a couple tests in 12cda1e53e8df5bd0428a75fc92b29bec1ee6900

1jeanpaul1 commented on CommunityFridgeMapApi/dependencies/python/storage.py:
```diff
@@ -0,0 +1,85 @@
+    return url
+
+
+class Storage:
```
Should this be called S3Storage? Or something S3 related?

1jeanpaul1 replied:
same common on the file name

0916dhkim replied:
The idea was to let lambda handlers persist files without worrying about the underlying implementation. It made sense when I was trying to save files in filesystem for local dev environments, but now it makes less sense with localstack for dev.

Renamed `Storage` to `S3Service` in 1e10fcb6a8406982905bd8000567a4b19faf433b.

0916dhkim commented:
> But when I go to the link a file is returned that I am not able to open. The file size is also very small (49 bytes) compared to the image I uploaded so I assume that's part of the problem

@1jeanpaul1 Sorry there was a typo in my curl command in PR description. I should have used `--data-binary` instead of `--data`. The updated command in the PR description should work properly.

1jeanpaul1 commented on README.md:
```diff
@@ -39,7 +39,7 @@ Follow these steps to get Dynamodb running locally
-    1. `aws dynamodb scan --table-name fridge --endpoint-url http://localhost:8000`
+    1. `aws dynamodb scan --table-name fridge --endpoint-url http://localhost:4566`
 
 ## API
```
Can we add a section here for the Image API? The instructions you wrote on the PR should be enough :) 

1. Start local SAM API
    ```
    sam local start-api --parameter-overrides ParameterKey=Environment,ParameterValue=local --docker-network cfm-network
    ```
1. Upload image (replace `<file-path>` with your actual image path like `"@/home/user/Downloads/sample.webp"`)
    ```
    curl --request POST \
      --url http://localhost:3000/v1/image \
      --header 'Content-Type: image/webp' \
      --data-binary "@<file-path>"
    ```

0916dhkim replied:
Good call! Done in 95d4a93dcb45142784f5f4308429fe9e1dbad8ee :+1:

1jeanpaul1 commented on CommunityFridgeMapApi/functions/image/v1/app.py:
```diff
@@ -0,0 +1,91 @@
+                    "Access-Control-Allow-Origin": "*",
+                },
+                "body": json.dumps({
+                    "error": "Unexpected error prevented server from fulfilling request."
```
Can you rename `"error"` to `"message"`. That's how what we've been naming the error messages in the other apis

0916dhkim replied:
Thanks for catching that :+1: 
Fixed here: 5302d2d6ea5006b1d696f2ae01110f32182182de

1jeanpaul1 commented on CommunityFridgeMapApi/functions/image/v1/app.py:
```diff
@@ -0,0 +1,91 @@
+                    "Access-Control-Allow-Origin": "*",
+                },
+                "body": json.dumps({
+                    "error": "Request could not be understood due to incorrect syntax."
```
We should mention that the image type must be webp in this error message?

Maybe this:
`Request could not be understood due to incorrect syntax. Image type must be webp.`

1jeanpaul1 replied:
Also can we rename `"error"` to `"message"`. That's how what we've been naming the error messages in the other apis

0916dhkim replied:
Fixed: 5302d2d6ea5006b1d696f2ae01110f32182182de

1jeanpaul1 commented:
Works locally for me :). I'm deploying this branch to test. I have a feeling we might be missing some permissions but not sure! Will post new comment after deployment

1jeanpaul1 commented on CommunityFridgeMapApi/dependencies/python/s3_service.py:
```diff
@@ -0,0 +1,101 @@
+        """
+        try:
+            self._client.create_bucket(Bucket=bucket)
+        except:
```
It's dangerous catching all the Exceptions. It makes it hard to debug and can result in errors/bugs going untreated. I think it would be better if we caught specific exceptions. 

I got this error when uploading an image in the deployed lambda: `{"error": "Unexpected error prevented server from fulfilling request."}`
    * `curl --request POST   --url https://0bebq1ckx7.execute-api.us-east-1.amazonaws.com/Prod/v1/image   --header 'Content-Type: image/webp'   --data-binary "@/Users/jtorre-admin/Downloads/DriversLicense.webp"`

However `curl --request POST   --url https://0bebq1ckx7.execute-api.us-east-1.amazonaws.com/Prod/v1/image   --header 'Content-Type: image/webp'   --data-binary "@/Users/jtorre-admin/Downloads/DriversLicense.jpeg"` returned the expected response
    * `{"error": "Request could not be understood due to incorrect syntax."}`

1jeanpaul1 replied:
Or when it's a more ambiguous exception. We should log the exception so we can debug like this: https://github.com/CollectiveFocus/CFM_Backend/blob/dev/CommunityFridgeMapApi/dependencies/python/db.py#L115

0916dhkim replied:
Yes, it does sound like `template.yaml` misconfiguration. Do we need to specify s3 buckets in that file?
Added some logging here: 66e70f19a9453f874a179ea5559eff0a418e687e

1jeanpaul1 commented:
I deployed but am getting an error, but unable to debug. 

Can we add some logging when there's an exception and write more specific exceptions

Comments here: 
https://github.com/CollectiveFocus/CFM_Backend/pull/16#discussion_r957848262

CollectiveFocus commented on CommunityFridgeMapApi/dependencies/python/s3_service.py:
```diff
@@ -0,0 +1,106 @@
+                bucket: name of the bucket
+        """
+        try:
+            self._client.create_bucket(Bucket=bucket)
```
Was getting this error in the deployed lambda. Lets create the S3 bucket from the template.yml file and only call this function if `os.getenv("Environment")` is `local`
```
An error occurred (AccessDenied) when calling the CreateBucket operation: Anonymous access is forbidden for this operation
```

0916dhkim replied:
That is a good idea. I wish we could just do `samlocal deploy`, but it doesn't work due to the aforementioned lambda layer paywall. Fixed :+1:

CollectiveFocus commented on CommunityFridgeMapApi/template.yaml:
```diff
@@ -119,6 +122,34 @@ Resources:
+          Properties:
+            Path: /v1/image
+            Method: post
+
```
Let's add an S3 resource: 
```
  ## S3 bucket
  CommunityFridgeMapImages:
    Type: AWS::S3::Bucket
    Properties:
      BucketName: community-fridge-map-images
      CorsConfiguration:
        CorsRules:
        - AllowedHeaders:
            - "*"
          AllowedMethods:
            - GET
            - PUT
            - HEAD
          AllowedOrigins:
            - "*"
```

CollectiveFocus commented on CommunityFridgeMapApi/functions/image/v1/app.py:
```diff
@@ -0,0 +1,91 @@
+
+    @staticmethod
+    def lambda_handler(event: dict, s3: S3Service) -> dict:
+        bucket = "CommunityFridgeMapImages"
```
bucket needs to follow s3 naming convention. Lets name it `community-fridge-map-images` ?

CollectiveFocus commented:
Thanks for adding the logging! The current issue I am running into is:
`
[ERROR]	2022-08-31T05:52:27.584Z	5d8e1243-c9e9-49e8-b686-1314656b1a6a	An error occurred (AccessDenied) when calling the PutObject operation: Access Denied
`

we'll need to play around with the policies in the template.yml file a little bit more!

0916dhkim commented:
@CollectiveFocus I think we're very close! Made the following changes:
- Suppress bucket creation on prod https://github.com/CollectiveFocus/CFM_Backend/pull/16/commits/4c08acb7a22e36ba30465fb073301d609090e965
- Renamed bucket to follow s3 naming convention https://github.com/CollectiveFocus/CFM_Backend/pull/16/commits/366647c68ab9103e351faa0fc3f20281cdb43f20
- Updated `template.yaml` https://github.com/CollectiveFocus/CFM_Backend/pull/16/commits/243bee8ea02bf6b26d3e5bb16cf9569e852b2617

CollectiveFocus commented:
> @CollectiveFocus I think we're very close! Made the following changes:
> 
> * Suppress bucket creation on prod [4c08acb](https://github.com/CollectiveFocus/CFM_Backend/commit/4c08acb7a22e36ba30465fb073301d609090e965)
> * Renamed bucket to follow s3 naming convention [366647c](https://github.com/CollectiveFocus/CFM_Backend/commit/366647c68ab9103e351faa0fc3f20281cdb43f20)
> * Updated `template.yaml` [243bee8](https://github.com/CollectiveFocus/CFM_Backend/commit/243bee8ea02bf6b26d3e5bb16cf9569e852b2617)

Thanks for the updates! Deployed and here's the API: https://gkckljhcvc.execute-api.us-east-1.amazonaws.com/Prod/v1/image
`
[ERROR]	2022-09-01T05:09:00.743Z	3f2de1a5-e040-4a4c-a9df-ec00d259de61	An error occurred (AccessDenied) when calling the PutObject operation: Access Denied
`

Getting the above error, so I think we need to play around with the policy in `ImageFunction`

or the s3 bucket in `CommunityFridgeMapImages`

0916dhkim commented:
@CollectiveFocus Any suggestions on what to try?
I've removed the `Sub!` directive [here](https://github.com/CollectiveFocus/CFM_Backend/pull/16/commits/fa89952133922031febca7b3da14eafa914ae8f8), but I can't think of anything that would cause an error.
Maybe we should check the bucket policy of `community-fridge-map-images` on AWS dashboard?
https://stackoverflow.com/a/70483737

CollectiveFocus commented:
The issue was at the bucket policy level. Here's the deployed api if you want to test it out: https://yi7brxeo1l.execute-api.us-east-1.amazonaws.com/Prod/v1/image

Here are the updates I made to get it to work:

1. Added Bucket Policy
```
  ImageFunctionBucketPolicy:
    Type: AWS::S3::BucketPolicy
    Properties:
      Bucket: !Ref CommunityFridgeMapImages
      PolicyDocument:
        Statement:
          -
            Action:
              - s3:*
            Effect: Allow
            Resource:
              - !Sub arn:aws:s3:::${CommunityFridgeMapImages}
              - !Sub arn:aws:s3:::${CommunityFridgeMapImages}/*
            Principal:
              AWS:
                - '*'
```

2.  Added another `Statement` field and used `!Sub`

```
ImageFunction:
    Type: AWS::Serverless::Function
    Properties:
      CodeUri: functions/image/v1
      Handler: app.lambda_handler
      Runtime: python3.9
      FunctionName: ImageFunction
      Environment:
        Variables:
          Environment: !Ref Environment
      Layers:
        - !Ref CommonLayer
      Policies:
        - Version: '2012-10-17'
          Statement:
          - Effect: Allow
            Action:
            - logs:CreateLogGroup
            - logs:CreateLogStream
            - logs:PutLogEvents
            Resource: arn:aws:logs:*:*:*
          Statement:
          - Effect: Allow
            Action:
            - s3:PutObject
            - s3:GetObject
            Resource:
            - !Sub "arn:aws:s3:::${CommunityFridgeMapImages}/*"
            - !Sub "arn:aws:s3:::${CommunityFridgeMapImages}"
      Events:
        UploadImage:
          Type: Api # More info about API Event Source: https://github.com/awslabs/serverless-application-model/blob/master/versions/2016-10-31.md#api
          Properties:
            Path: /v1/image
            Method: post
```

Will want to make bucket policy stricter next month! I'll make a ticket

CollectiveFocus commented:
```
  ##########################
  ImageFunctionBucketPolicy:
    Type: AWS::S3::BucketPolicy
    Properties:
      Bucket: !Ref CommunityFridgeMapImages
      PolicyDocument:
        Statement:
          -
            Action:
              - s3:*
            Effect: Allow
            Resource:
              - !Sub arn:aws:s3:::${CommunityFridgeMapImages}
              - !Sub arn:aws:s3:::${CommunityFridgeMapImages}/*
            Principal:
              AWS:
                - '*'
```

Your indentation is off by one in`ImageFunctionBucketPolicy`

make sure `sam build --use-container` runs when you push

CollectiveFocus reviewed (approved):
Looks good! Thank you :)
